examples/tutorial/tutorial_notebooks/test/test_notebooks.py

```diff
@@ -0,0 +1,188 @@
+def test_exercise_1():
+    modified_notebook_path = (
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_1_test.ipynb"
+    )
+    nb = nbformat.read(
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_1.ipynb",
+        as_version=nbformat.NO_CONVERT,
+    )
+    _replace_str(nb, "import pandas as pd", "import modin.pandas as pd")
+
+    nbformat.write(nb, modified_notebook_path)
+    _execute_notebook(modified_notebook_path)
+
+
+# this notebook works "as is" but for testing purposes we can use smaller dataset
+def test_exercise_2():
+    modified_notebook_path = (
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_2_test.ipynb"
+    )
+    nb = nbformat.read(
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_2.ipynb",
+        as_version=nbformat.NO_CONVERT,
+    )
+
+    _replace_str(
+        nb,
+        'path = "s3://dask-data/nyc-taxi/2015/yellow_tripdata_2015-01.csv"',
+        '# path = "s3://dask-data/nyc-taxi/2015/yellow_tripdata_2015-01.csv"',
+    )
+
+    new_optional_cell = 'path = "taxi.csv"\n' + download_taxi_dataset
+
+    optional_cell_idx = _find_code_cell_idx(nb, "[Optional] Download data locally.")
+    nb["cells"][optional_cell_idx]["source"] = new_optional_cell
+
+    nbformat.write(nb, modified_notebook_path)
+    _execute_notebook(modified_notebook_path)
+
+
+# in this notebook user should add custom mad implementation
+# to make notebook work
+def test_exercise_3():
+    modified_notebook_path = (
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_3_test.ipynb"
+    )
+    nb = nbformat.read(
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_3.ipynb",
+        as_version=nbformat.NO_CONVERT,
+    )
+
+    user_mad_implementation = """PandasQueryCompiler.sq_mad_custom = TreeReduce.register(lambda cell_value, **kwargs: cell_value ** 2,
+                                                             pandas.DataFrame.mad)
+
+def sq_mad_func(self, axis=None, skipna=True, level=None, **kwargs):
+    if axis is None:
+        axis = 0
+
+    return self._reduce_dimension(
+        self._query_compiler.sq_mad_custom(
+            axis=axis, skipna=skipna, level=level, **kwargs
+        )
+    )
+
+pd.DataFrame.sq_mad_custom = sq_mad_func
+
+modin_mad_custom = df.sq_mad_custom()
+    """
+
+    _replace_str(nb, "modin_mad_custom = ...", user_mad_implementation)
+
+    nbformat.write(nb, modified_notebook_path)
+    _execute_notebook(modified_notebook_path)
+
+
+# this notebook works "as is" but for testing purposes we can use smaller dataset
+def test_exercise_4():
+    modified_notebook_path = (
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_4_test.ipynb"
+    )
+    nb = nbformat.read(
+        "examples/tutorial/tutorial_notebooks/introduction/exercise_4.ipynb",
+        as_version=nbformat.NO_CONVERT,
+    )
+
+    s3_path_cell = 's3_path = "taxi.csv"\n' + download_taxi_dataset
```

Comment:
What is the dataset here?

Reply:
It is `https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2021-01.csv` dataset. It is downloaded only once and then cached with `actions/cache@v2`.

Reply:
Oh, I see.

Reply:
Should we add a variable for "taxi.csv" at the header of this file? Just in order to not forget about this string if we change it in some place.